base/inference.jl

```diff
@@ -791,23 +791,36 @@ function getfield_tfunc(s00::ANY, name)
 end
 add_tfunc(getfield, 2, 2, (s::ANY, name::ANY) -> getfield_tfunc(s, name))
 add_tfunc(setfield!, 3, 3, (o::ANY, f::ANY, v::ANY) -> v)
-function fieldtype_tfunc(s::ANY, name::ANY)
-    if isa(s, Const)
-        s = s.val
-    elseif isType(s)
-        s = s.parameters[1]
-    else
-        return Type
+function fieldtype_tfunc(s0::ANY, name::ANY)
+    s = instanceof_tfunc(s0)
+    exact = (isa(s0, Const) || isType(s0)) && !has_free_typevars(s)
+    if isa(name, Const)
+        fld = name.val
+        if !isa(fld, Int) && !isa(fld, Symbol)
+            return Bottom
+        end
+        if exact
+            ft = nothing
+            try
+                ft = fieldtype(s, fld)
+            catch
+            end
+            if ft !== nothing
+                if isleaftype(ft) || ft === Bottom
+                    return Const(ft)
+                else
+                    return Type{ft}
```

Comment:
This seems like it might not be quite monotonic, since I think the `getfield_tfunc` is actually pretty smart about inferring `Const` for various cases. For example, the return_type of `HasAbstractlyTypedField` should be `Const(ft)` (although we widen it before letting the user see that, so it's a bit annoying to test for). It should be possible to see the difference by using a UnionAll type for the field, and checking that you can parameterize the result of fieldtype to form a leaftype.

Also, would be nice not to catch errors silently in that try/catch.